Meta/serenity.sh

```diff
@@ -0,0 +1,190 @@
+shift
+if [ $CMD = "help" ]; then
+  print_help
+  exit 0
+fi
+
+if [ ! -z "$1" ]; then
+  TARGET="$1"; shift
+else
+  TARGET="i686"
+fi
+CMD_ARGS=( "$@" )
+CMAKE_ARGS=()
+
+function get_top_dir()
+{
+  git rev-parse --show-toplevel
+}
+
+function is_valid_target()
+{
+  if [ $TARGET = "lagom" ]; then
+    echo 'target is lagom';
+    CMAKE_ARGS+=("-DBUILD_LAGOM=ON")
+    return 0
+  fi
+  [[ $TARGET =~ ^(i686|x86_64|lagom)$ ]] || return 1
+}
+
+function create_build_dir()
+{
+  mkdir -p "$BUILD_DIR"
+  ( cd "$BUILD_DIR" && cmake -GNinja "${CMAKE_ARGS[@]}" ../..)
+}
+
+function cmd_with_target()
+{
+  is_valid_target || ( >&2 echo "Unknown target: $TARGET"; usage )
+  export SERENITY_ARCH="$TARGET"
+  SERENITY_ROOT=$(get_top_dir)
+  export SERENITY_ROOT
+  BUILD_DIR="$SERENITY_ROOT/Build/$TARGET"
+  if [ $TARGET = "lagom" ]; then
+    TOOLCHAIN=i686
+  else
+    TOOLCHAIN="$TARGET"
+  fi
+  TOOLCHAIN_DIR="$SERENITY_ROOT/Toolchain/Build/$TARGET"
+}
+
+function ensure_target()
+{
+  [ -d "$BUILD_DIR" ] || create_build_dir
+}
+
+function build_target()
+{
+  ( cd "$BUILD_DIR" && ninja "$@" )
+}
+
+function delete_target()
+{
+  [ ! -d "$BUILD_DIR" ] || rm -rf "$BUILD_DIR"
+}
+
+function build_toolchain()
+{
+  ( cd Toolchain && ARCH="$TOOLCHAIN" ./BuildIt.sh )
+}
+
+function ensure_toolchain()
+{
+  [ -d "$TOOLCHAIN_DIR" ] || build_toolchain
+}
+
+function delete_toolchain()
+{
+  [ ! -d "$TOOLCHAIN_DIR" ] || rm -rf "$TOOLCHAIN_DIR"
+}
+
+function not_supported_lagom()
+{
+  >&2 echo "Command '$CMD' not supported for the lagom target"
+  exit 1
+}
+
+if [[ $CMD =~ ^(build|install|image|run|rebuild|recreate|addr2line|setup-and-run)$ ]]; then
+  cmd_with_target
+  [ $CMD != "recreate" -a $CMD != "rebuild" ] || delete_target
+  ensure_target
+  ensure_toolchain
+  case "$CMD" in
+    build)
+      build_target "$@"
+      ;;
+    install)
+      [ $TARGET != "lagom" ] || not_supported_lagom
+      build_target
+      build_target install
+      ;;
+    image)
+      [ $TARGET != "lagom" ] || not_supported_lagom
+      build_target
+      build_target install
+      build_target image
+      ;;
+    run)
+      build_target
+      if [ $TARGET = "lagom" ]; then
+        build_target test
+      else
+        build_target install
+        build_target image
+        if [ ! -z "${CMD_ARGS[0]}" ]; then
+          export SERENITY_KERNEL_CMDLINE="${CMD_ARGS[0]}"
+        fi
+        build_target run
+      fi
+      ;;
+    rebuild)
+      build_target "$@"
+      ;;
+    recreate)
+      ;;
+    addr2line)
+      build_target
+      [ $# -ge 2 ] || usage
+      BINARY_FILE="$1"; shift
+      "$TOOLCHAIN_DIR/binutils/binutils/addr2line" -e "$BUILD_DIR/$BINARY_FILE" "$@"
+      ;;
+    *)
+      build_target "$CMD" "$@"
+      ;;
+  esac
+elif [ $CMD = "delete" ]; then
+  cmd_with_target
+  delete_target
+elif [ $CMD = "rebuild-toolchain" ]; then
+  cmd_with_target
+  [ "$TARGET" != "lagom" ] || ( >&2 echo "The lagom target uses the host toolchain"; exit 1 )
```

Comment:
Running `exit 1` inside `()` is basically the same thing as running `false`; this is probably what you want instead.

```suggestion
  [ "$TARGET" != "lagom" ] || { echo "The lagom target uses the host toolchain" >&2; exit 1 ;}
```

I think it would be better to just create a `die` helper function that prints an error message and exits like we do in all the other shell scripts though.

Also please indent with 4 spaces and define functions as `function_name() {\ncode\n}` since that's the code style we are already using for shell scripts.

Reply:
I think it will also work as you wrote it, but only because of `set -e`.

Reply:
Doesn't this apply to `/bin/false` either? Only works due to `set -e`? I'm not a shell expert, if `false` is the right thing to use I'm happy to change it.

Reply:
No, the correct thing would be to use `{}` instead of `()`.
`()` runs commands inside a subshell and, if you run `exit 1` in the subshell, you will only exit (with exit status 1) from the subshell not the main script (hence equivalent to running `false`). Since `set -e` is activated and `(...; exit 1)` returns a non-0 exit status, the main script will exit anyway, but only because the script automatically kills itself when a command returns non-0 and `set -e` is on, not because `exit` actually made it exit.
`{}` just groups commands and runs them in the current shell (not in a subshell), that is probably what you want here.

Reply:
But I think it's better to add an helper `die` function like the one that can be found in `Meta/run.sh` (and other shell scripts):
```
die() {
    echo "die: $*"
    exit 1
}
```

And then use that:

```suggestion
  [ "$TARGET" != "lagom" ] || die 'The lagom target uses the host toolchain'
```

Reply:
https://github.com/SerenityOS/serenity/pull/5686#discussion_r589078701

> Also please indent with 4 spaces and define functions as `function_name() {\ncode\n}` since that's the code style we are already using for shell scripts.

@tomuta, I think you also missed this part. :^)

Reply:
I did miss this. Now if github would only expand those collapsed resolved conversations when comments are added... ;)